sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -483,6 +484,25 @@ object CollapseRepartition extends Rule[LogicalPlan] {
 }
 
 /**
+ * Collapse two adjacent [[Sort]] operators into one if possible. Keep the last sort
+ * This rule applies to the scenario where the global is same for the Sort nodes and then
+ * either a) The sorts are adjacent or b) In between two Sort nodes, there is a Filter or
+ * a Project or a Limit
+ */
+object CollapseSorts extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
```

Comment:
as it is, i'm not sure how useful this rule is.

I think what would be useful is to have an EliminateSort rule, that can actually eliminate the more useful cases.